catalog/plugin/catalog-plugin-duplication/src/main/java/ddf/catalog/metacard/duplication/DuplicationValidator.java

```diff
@@ -0,0 +1,287 @@
+                    .collect(Collectors.toList());
+
+            String message = String.format("Duplicate data found in catalog for ID {%s}.",
+                    metacard.getId());
+            final ValidationExceptionImpl exception = new ValidationExceptionImpl(message);
+            exception.setErrors(errors);
+            exception.setWarnings(warnings);
+            throw exception;
+        }
+    }
+
+    private Set<ValidationViolation> reportDuplicates(final Metacard metacard) {
+
+        Set<ValidationViolation> violations = new HashSet<>();
+
+        if (ArrayUtils.isNotEmpty(warnOnDuplicateAttributes)) {
+            ValidationViolation warnValidation = reportDuplicates(metacard,
+                    warnOnDuplicateAttributes,
+                    ValidationViolation.Severity.WARNING);
+            if (warnValidation != null) {
+                violations.add(warnValidation);
+            }
+        }
+        if (ArrayUtils.isNotEmpty(errorOnDuplicateAttributes)) {
+            ValidationViolation errorViolation = reportDuplicates(metacard,
+                    errorOnDuplicateAttributes,
+                    ValidationViolation.Severity.ERROR);
+            if (errorViolation != null) {
+                violations.add(errorViolation);
+            }
+        }
+
+        return violations;
+
+    }
+
+    private ValidationViolation reportDuplicates(final Metacard metacard, String[] attributeNames,
+            ValidationViolation.Severity severity) {
+
+        Set<String> duplicates = new HashSet<>();
+        ValidationViolation violation = null;
+
+        final Set<String> uniqueAttributeNames = Stream.of(attributeNames)
+                .filter(attribute -> metacard.getAttribute(attribute) != null)
+                .collect(Collectors.toSet());
+        final Set<Attribute> uniqueAttributes = uniqueAttributeNames.stream()
+                .map(attribute -> metacard.getAttribute(attribute))
+                .collect(Collectors.toSet());
+        if (!uniqueAttributes.isEmpty()) {
+            LOGGER.debug("Checking for duplicates for id {} against attributes [{}]",
+                    metacard.getId(),
+                    collectionToString(uniqueAttributeNames));
+
+            SourceResponse response = query(uniqueAttributes);
+            if (response != null) {
+                response.getResults()
+                        .forEach(result -> duplicates.add(result.getMetacard()
+                                .getId()));
+            }
+            if (!duplicates.isEmpty()) {
+
+                violation = createViolation(uniqueAttributeNames, duplicates, severity);
+                LOGGER.debug(violation.getMessage());
+
+                return violation;
+            }
+        }
+        return violation;
+    }
+
+    private Filter[] buildFilters(Set<Attribute> attributes) {
+        List<Filter> filters = new ArrayList<>();
+        attributes.forEach(attribute -> {
```

Comment:
This should probably be a couple of `stream.map()`s instead of a `forEach()` and then you can collect to an array and just return.